flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/DeltaJoinUtil.java

```diff
@@ -0,0 +1,407 @@
+     */
+    private static final Set<FlinkJoinType> ALL_SUPPORTED_JOIN_TYPES =
+            Sets.newHashSet(FlinkJoinType.INNER);
+
+    private DeltaJoinUtil() {}
+
+    /** Check whether the {@link StreamPhysicalJoin} can be optimized into a delta join. */
+    public static boolean canConvertToDeltaJoin(StreamPhysicalJoin join) {
+        if (!isJoinTypeSupported(join)) {
+            return false;
+        }
+
+        if (!areJoinConditionsSupported(join)) {
+            return false;
+        }
+
+        // delta join with eventual consistency will send duplicate changes to downstream nodes
+        if (!canJoinOutputDuplicateChanges(join)) {
+            return false;
+        }
+
+        // currently, only join with append-only inputs is supported
+        if (!areAllInputsInsertOnly(join)) {
+            return false;
+        }
+
+        if (!areAllJoinInputsInWhiteList(join)) {
+            return false;
+        }
+
+        return areAllJoinTableScansSupported(join);
+    }
+
+    /**
+     * Get the {@link RelOptTable} from the {@link TableScan} recursively on the input of this node.
+     */
+    public static RelOptTable getTableScanRelOptTable(RelNode node) {
+        return getTableScan(node).getTable();
+    }
+
+    /**
+     * Extract the delta join spec used for {@link StreamPhysicalDeltaJoin} from {@link
+     * StreamPhysicalJoin}.
+     */
+    public static DeltaJoinSpec getDeltaJoinSpec(
+            StreamPhysicalJoin join, boolean treatRightAsLookupSide) {
+        JoinInfo joinInfo = join.analyzeCondition();
+        RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+
+        RexNode condition = RexUtil.composeConjunction(rexBuilder, joinInfo.nonEquiConditions);
+        Optional<RexNode> remainingCondition =
+                condition.isAlwaysTrue() ? Optional.empty() : Optional.of(condition);
+
+        final RelOptTable lookupRelOptTable;
+        List<IntPair> streamToLookupJoinKeys = joinInfo.pairs();
+        if (treatRightAsLookupSide) {
+            lookupRelOptTable = DeltaJoinUtil.getTableScanRelOptTable(join.getRight());
+        } else {
+            streamToLookupJoinKeys = reverseIntPairs(streamToLookupJoinKeys);
+            lookupRelOptTable = DeltaJoinUtil.getTableScanRelOptTable(join.getLeft());
+        }
+        Preconditions.checkState(lookupRelOptTable instanceof TableSourceTable);
+        final TableSourceTable lookupTable = (TableSourceTable) lookupRelOptTable;
+
+        Map<Integer, FunctionCallUtils.FunctionParam> allLookupKeys =
+                analyzerDeltaJoinLookupKeys(streamToLookupJoinKeys);
+
+        return new DeltaJoinSpec(
+                new TemporalTableSourceSpec(lookupTable),
+                allLookupKeys,
+                remainingCondition.orElse(null));
+    }
+
+    /** Check if the join type is supported during delta join optimization. */
+    public static void validateSupportedJoinType(FlinkJoinType joinType) {
```

Comment:
use `isJoinTypeSupported` instead and add `Preconditions` to throw unexpected exception in exec itself